macos/FluentUI/Button/Button.swift

```diff
@@ -175,304 +234,402 @@ open class Button: NSButton {
+				return
+			}
+			setSizeParameters(forSize: size)
+			needsDisplay = true
 		}
 	}
-	
-	private var hoverBackgroundColor: NSColor {
-		return restBackgroundColor.withSystemEffect(.rollover)
-	}
+}
 
-	private var pressedBackgroundColor: NSColor {
-		return restBackgroundColor.withSystemEffect(.pressed)
-	}
+// MARK: - ButtonCell
 
-	private var layerBackgroundColor: NSColor {
-		if isEnabled {
-			if mouseDown {
-				return pressedBackgroundColor
-			} else if mouseEntered {
-				return hoverBackgroundColor
-			} else {
-				return restBackgroundColor
-			}
-		} else {
-			return disabledColor(for: restBackgroundColor)
+class ButtonCell: NSButtonCell {
+	fileprivate var verticalPadding: CGFloat = 0
+	fileprivate var horizontalPadding: CGFloat = 0
+	fileprivate var titleVerticalPositionAdjustment: CGFloat = 0
+	fileprivate var titleToImageSpacing: CGFloat = 0
+	fileprivate var titleToImageVerticalSpacingAdjustment: CGFloat = 0
+
+	override func imageRect(forBounds rect: NSRect) -> NSRect {
+		guard let image = image else {
+			return NSZeroRect
+		}
+
+		let isRTL = NSApp.userInterfaceLayoutDirection == .rightToLeft
+		let imageRect = super.imageRect(forBounds: rect)
+		let titleSize = title.size(withAttributes: [.font: font as Any])
+		let imageSize = image.size
+
+		var x = imageRect.origin.x
+		var y = imageRect.origin.y
+		var width = imageRect.size.width
+		var height = imageRect.size.height
+
+		switch imagePosition {
+		case .noImage:
+			x = 0
+			y = 0
+			width = 0
+			height = 0
+		case .imageOnly, .imageOverlaps:
+			x = horizontalPadding
+			y = verticalPadding
+			width = rect.width - (horizontalPadding * 2)
+			height = rect.height - (verticalPadding * 2)
+		case .imageLeft:
+			x = horizontalPadding
+			y = (rect.height - imageSize.height) / 2
+			width = imageSize.width
+			height = imageSize.height
+		case .imageLeading:
+			x = isRTL ? titleSize.width + titleToImageSpacing + horizontalPadding : horizontalPadding
+			y = (rect.height - imageSize.height) / 2
+			width = imageSize.width
+			height = imageSize.height
+		case .imageRight:
+			x = titleSize.width + titleToImageSpacing + horizontalPadding
+			y = (rect.height - imageSize.height) / 2
+			width = imageSize.width
+			height = imageSize.height
+		case .imageTrailing:
+			x = isRTL ? horizontalPadding : titleSize.width + titleToImageSpacing + horizontalPadding
+			y = (rect.height - imageSize.height) / 2
+			width = imageSize.width
+			height = imageSize.height
+		case .imageBelow:
+			x = (rect.width - imageSize.width) / 2
+			y = (rect.height - imageSize.height + titleSize.height + titleToImageSpacing - titleToImageVerticalSpacingAdjustment) / 2
+			width = imageSize.width
+			height = imageSize.height
+		case .imageAbove:
+			x = (rect.width - imageSize.width) / 2
+			y = (rect.height - imageSize.height - titleSize.height - titleToImageSpacing + titleToImageVerticalSpacingAdjustment) / 2
+			width = imageSize.width
+			height = imageSize.height
+		@unknown default:
```

Comment:
can we delete default?

Reply:
This default cases are in the original code.  I can remove them.

Reply:
JK removing this breaks the build.  Xcode gives me this error and has an auto-fix for it:

Switch covers known cases, but 'NSControl.ImagePosition' may have additional unknown values, possibly added in future versions

Handle unknown values using "@unknown default"

Reply:
Yeah, I'm not sure why this switch needs that `@uknown` param but others don't. Probably something about it coming from objective c.

Reply:
Yeah swift 5 added a feature for this which is a regression imo. See https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/grouping_related_objective-c_constants for some context, but essentially because this is not marked as NS_CLOSED_ENUM it is not frozen and thus may add values in the future. As a "convenience", the swift compiler forces you to add unknown default for this so your code automatically still compiles even when new enums get added.

I might be missing a subtlety of it, but to me that negates one of the main benefits of using switch/case statements in that you can avoid the default case and get a compiler error if a new enum value is added, requiring each case to go explicitly decide how to handle this new case. I believe adding a new case would now result in a warning rather than an error if an unknown default case exists, but don't get me started on the many flaws of SWIFT_TREAT_WARNINGS_AS_ERRORS :). I personally would vote to use NS_CLOSED_ENUM wherever we have code that may be consumed by swift.